tag-expressions-python: Prepare release v4.0.1

jenisys opened the pull request:
## Summary

Perform some cleanups and updates before a new release of this package can be published on https://pypi.org .

## Details

* FIXED: Out-of-date version info in some places
* UPDATED: `tox.ini` to use newer Python versions
* README: Use badge for CircleCI build server status (used in published version on pypi.org; was still using Travis CI)

PATCH RELEASE CONTAINS:

* FIX #1736 (with pull #1742 ; required by: `setuptools >= 58.0.2`)

## Motivation and Context

Prepare for next release of `tag-expressions-python`.

## How Has This Been Tested?

Normal test suite was run with tox.
BUT: Only cosmetic changes were performed here.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue) -- Minor cosmetic changes to next prepare release
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] The change has been ported to Python.
- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the CHANGELOG accordingly.

mpkorstanje commented:
I see that a release has been made for python but not the other implementations.

All implementations share a version number. To avoid problems in the future can you make a release of the other implementations too?

jenisys commented:
@mpkorstanje 
Have you looked at the changes and that I created a patch version?

mpkorstanje commented:
Even if a patch only affects one implementation all implementations get a new release. This helps keep things simple. As we don't have to wonder about the available version numbers for each version.

If you use the make file it does not take any extra effort. 

https://github.com/cucumber/common/blob/main/RELEASE_PROCESS.md

Can I assume you did not use the make file?

aslakhellesoy commented:
I second what @mpkorstanje said. Following the release process also creates a tag in git, and updates the changelog with links to the released version and adds a new section for the next version. Right now it is not obvious in the changelog and git tags that a 4.0.1 release has been made.

aurelien-reeves commented:
@jenisys It is nice to see you back there!

Thank you for your contribution. As you may have noticed, python implementation of `cucumber` lacked some activity the past couple of years.

The python package seems not to have been actually released, but simply pre-released as the name of the PR suggest: https://pypi.org/project/cucumber-tag-expressions/#history

Could you confirm?

@mpkorstanje where did you see a release for that module?

When I check your changes, it looks like the Makefile for python are missing some places where to update the version number: `.bumpversion.cfg`, `__init__.py`, and `pytest.ini`. We have to fix this :p


And please, don't worry for that release thing. If any mistake has been made, that's not a big deal. It happens to everybody. Do not be afraid to contribute again!

jenisys commented:
I released the Python packages but noticed that there is still a flaw in the README that is used for the package description (I fixed the flaw but I have not pushed it back to the repo). Therefore, I removed the release afterwards. Due to this removal, the package cannot be republished with version `4.0.1` due to some pypi constraints. It needs to have another version number, like `4.0.2`.

NOTES:
* I tried to prepare the release of the tag-expression but I stumbled over a problem with `tag-expression/javascript` package related to the `npm install` step where it was complaining that a package `cucumber/tag-expressions@v4.0.1` was missing (that was prepared in a pre-release just yet).
* I did not change the Makefile(s). On first inspection, I am also not aware what I should change in there (because I have not found any version related info in them).

aurelien-reeves commented:
tag-expressions has been released as 4.0.2

The `npm install` issue has been fixed cleaning the node modules (`npm run clean`). I've updated the release process documentation accordingly.

aslakhellesoy commented:
This will be a lot simpler once we move the code to a new `cucumber/tag-expressions` repo (see #1724)

aurelien-reeves commented:
Yes, I guess so

Once again @jenisys, thanks for your contribution! I hope to see you more often to take care of the python port! 🤩

aslakhellesoy commented:
Hi @jenisys - completely unrelated to this - would you be interested in trying to grow the behave contributor community? @aurelien-reeves and I have been talking to @bbelderbos and @blaisep about coaching the pybit.es community towards making contributions to behave :-)
